File: 2023/python/day02_cube_contundrum/part1.py
import io
import itertools
from collections import defaultdict
from typing import NamedTuple

BAG = defaultdict(
    int,
    {
        "red": 12,
        "green": 13,
        "blue": 14,
    },
)


class Record(NamedTuple):
    count: int
    color: str


def parse_record(s: str) -> Record:
    count, color = s.split(" ")
    return Record(int(count), color)


def parse_set(s: str):
    return list(map(parse_record, s.split(", ")))


def parse_sets(s: str):
    return list(map(parse_set, s.split("; ")))


def parse_game(s: str):
    idx = s.find(":")
    game_id = s[5:idx]
    records = s[idx + 2 :].strip()
    return int(game_id), parse_sets(records)


def process_game(line: str) -> int:
    game_id, game = parse_game(line)
    is_game_playable = all(
        count <= BAG[color] for count, color in itertools.chain.from_iterable(game)
    )
    if is_game_playable:
        return game_id
    return 0


def cube_conundrum(data: io.TextIOWrapper) -> int:
    return sum(map(process_game, data))

File: lib/day02/part1.ex
defmodule Aoc.Day02part1 do
  @spec main(binary) :: integer
  def main(data) do
    parse_record = fn s ->
      [count, color] = String.split(s)
      {String.to_integer(count), color}
    end

    parse_set = fn s ->
      s
      |> String.split(", ")
      |> Enum.map(parse_record)
    end

    parse_game = fn s ->
      [part1, part2] = String.split(s, ": ")
      ["Game", id] = String.split(part1)

      sets =
        part2
        |> String.split("; ")
        |> Enum.map(parse_set)

      {String.to_integer(id), sets}
    end

    check_in_bag = fn
      {count, color} ->
        case color do
          "red" -> count <= 12
          "green" -> count <= 13
          "blue" -> count <= 14
        end
    end

    process_game = fn
      {game_id, game} ->
        is_playable =
          Enum.all?(game, fn set ->
            set
            |> Enum.map(check_in_bag)
            |> Enum.all?()
          end)

        if is_playable do
          game_id
        else
          0
        end
    end

    data
    |> String.split("\n")
    |> Enum.filter(fn s -> s != "" end)
    |> Enum.map(parse_game)
    |> Enum.map(process_game)
    |> Enum.sum()
  end
end
